### PR TITLE
planner: Fix Join reorder occurs "index out of range" error

### LINKED ISSUE #24095: Join reorder occurs "index out of range" error
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

```
mysql> create table t (id int, value decimal(10,5));
Query OK, 0 rows affected (0.01 sec)

mysql> desc format = 'brief' select count(*) from t join (select t.id, t.value v1 from t join t t1 on t.id = t1.id order by t.value limit 1) v on v.id = t.id and v.v1 = t.value;
ERROR 1105 (HY000): runtime error: index out of range [4] with length 4
```

### 2. What did you expect to see? (Required)
Empty result

### 3. What did you see instead (Required)
Error

### 4. What is your TiDB version? (Required)
Master (873fbfd0871462c770e750be4c4e3e5acf01fddd)

### PR DESCRIPTION
**Pull request by mmyj**
### What problem does this PR solve?

Issue Number: close #24095 

Problem Summary:

```
for i, col := range p.Schema().Columns {
  if !col.Equal(nil, originalSchema.Columns[i]) {
	  schemaChanged = true
	  break
  }
}
```
Panic when `len(p.Schema().Columns) != len(originalSchema.Columns)`

### Check List 

Tests 

- Unit test

### Release note 
- Fix Join reorder occurs "index out of range" error

### CODE REVIEW

File: planner/core/rule_join_reorder_test.go [hunk 1]
```diff
@@ -0,0 +1,41 @@
+// Copyright 2021 PingCAP, Inc.
```
Comment (hanfei1991) [inline planner/core/rule_join_reorder_test.go:1]: Good job! But indeed this test is an integrated test. How about moving it to "integration test" and check the result of explain?

Comment (mmyj) [reply planner/core/rule_join_reorder_test.go]: done

**Review by hanfei1991 — APPROVED** [review]

**Review by eurekaka — COMMENTED** [review]
/lgtm

**Pull request merged**